Adding Technical docs (#1)
xdaluca/erc-7786: docs/pages/examples.mdx

# Examples

> ## 1. Sending Message using ERC7786 with Axelar adapter

Reviewer: ```suggestion
## 1. Sending a message using ERC7786 with Axelar adapter
```

This tutorial is using `Base Sepolia` as source chain and `Arbitrum Sepolia` as destination chain.

### Concepts

ERC7786 defines two main interfaces for cross-chain messaging:

1. `IERC7786GatewaySource` - The source chain interface that allows sending messages. Key components:
   - `sendMessage`: Function to initiate message sending, taking:

>      - destinationChain (CAIP-2 format, e.g. "eip155:421614" for Arbitrum Sepolia)

Reviewer: ```suggestion
     - destinationChain (CAIP-2 format, e.g. `eip155:421614` for Arbitrum Sepolia)
```

     - receiver address
     - payload (arbitrary bytes)
     - attributes (optional protocol-specific parameters)
   - Returns an outbox ID to track the message
   - Emits a `MessagePosted` event

2. `IERC7786Receiver` - The destination chain interface that receives messages. Key components:
   - `executeMessage`: Function called by the destination gateway to deliver the message
   - Must verify the calling gateway is trusted

>    - Must return a specific selector (0x675b049b) to confirm successful execution

Reviewer: ```suggestion
   - Must return a specific selector (`0x675b049b`) to confirm successful execution
```

Important concepts to understand:

- Messages require proper setup of source and destination gateways that implement these interfaces

> - Chain identifiers follow the CAIP-2 standard (e.g. "eip155:421614" for Arbitrum Sepolia)

Reviewer: This is repeating too closely the line above.

- Some protocols (like Axelar) may require additional steps like:
  - Registering chain equivalences between protocol-specific and CAIP-2 identifiers
  - Registering remote adapter addresses
  - Paying for destination chain gas fees and potentially relayer or protocol fees
  - Message approval and execution on the destination chain

The standard is designed to be protocol-agnostic while allowing protocol-specific features through the attributes parameter.

> ### Setup

Reviewer: Eventually there should be a directory of predeployed adapters. Users shouldn't be the ones doing this setup. So I'd try to clarify this in some way. Maybe changing the section title to "Adapter Setup".

Author: It makes sense. I'll just change the section title to "Adapter Setup" as you suggested for now.

both directions, we deployed [`AxelarGatewayDuplex.sol`](https://github.com/OpenZeppelin/openzeppelin-community-contracts/blob/master/contracts/crosschain/axelar/AxelarGatewayDuplex.sol) in both chains.
Note that when deploying the contract, one of the constructor parameters is the Axelar gateway proxy, which is the contract responsible for receiving (or sending) all messages in the respective chain. 
Note that each chain has its own gateway proxy (at least one). This archicteture is used to centralize all messages in one place, where relayers can easily listen for events. 
This is equivalent to LayerZero Endpoints and Hyperlane Mailboxes, for example. The contract addresses for the gateways on different chains for Axelar can be seen at its [docs](https://docs.axelar.dev/resources/contract-addresses/testnet/).
    
    - Contract deployed on source chain: [0xBb6A11d2FdEF18522cC9d14DdAdb676c8c126AEE](https://sepolia.basescan.org/address/0xBb6A11d2FdEF18522cC9d14DdAdb676c8c126AEE)
    - Contract deployed on destination chain: [0x050193203845D4145Dea5200C7eA49EF28b5f870](https://sepolia.arbiscan.io/address/0x050193203845d4145dea5200c7ea49ef28b5f870)

> 2. With the contracts deployed (source and destination chains), we need to register the chain equivalence between axelar ids and the ERC7786 standard. Note that when sending a message, the destination chain [parameter](https://github.com/OpenZeppelin/openzeppelin-community-contracts/blob/cc554fc5cfee7498dc0177beb01118e4499826f6/contracts/crosschain/axelar/AxelarGatewaySource.sol#L29) 

Reviewer: ```suggestion
2. With the contracts deployed (source and destination chains), we need to register the chain equivalence between axelar ids and the ERC7786 standard. This is because when sending a message, the destination chain [parameter](https://github.com/OpenZeppelin/openzeppelin-community-contracts/blob/cc554fc5cfee7498dc0177beb01118e4499826f6/contracts/crosschain/axelar/AxelarGatewaySource.sol#L29) 
```

should follow the [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-2.md) standard to identify the destination chain. However, Axelar uses its own set of ids for the [chains](https://docs.axelar.dev/resources/contract-addresses/testnet/). This is also the case for other protocols such as LayerZero, with its [eids](https://docs.layerzero.network/v2/deployments/deployed-contracts).

> This way, it's important to map CAIP-2 identifiers to each bridge protocol set of chain identifiers. [`AxelarGatewatBase.sol`](https://github.com/OpenZeppelin/openzeppelin-community-contracts/blob/cc554fc5cfee7498dc0177beb01118e4499826f6/contracts/crosschain/axelar/AxelarGatewayBase.sol#L52) solves this by registering a chain equivalence with the identifiers.

Reviewer: ```suggestion
This makes it necessary to keep a mapping of CAIP-2 identifiers to each bridge protocol's own set of chain identifiers. [`AxelarGatewatBase.sol`](https://github.com/OpenZeppelin/openzeppelin-community-contracts/blob/cc554fc5cfee7498dc0177beb01118e4499826f6/contracts/crosschain/axelar/AxelarGatewayBase.sol#L52) solves this by registering a chain equivalence with the identifiers.
```
